Add blog post about CRI-O's latest packaging efforts (#43128)
kubernetes/website: content/en/blog/_posts/2023-10-10-cri-o-community-package-infrastructure.md

Reviewer: Early feedback: site relative links please

ie `/blog/2023/08/31/legacy-package-repository-deprecation/` not `/blog/2023/08/31/legacy-package-repository-deprecation/`

Reviewer 2: Tanks

---
layout: blog
title: "CRI-O is moving towards pkgs.k8s.io"
date: 2023-10-10
slug: cri-o-community-package-infrastructure
---

**Author:** Sascha Grunert

> The Kubernetes community [recently announced](https://kubernetes.io/blog/2023/08/31/legacy-package-repository-deprecation/)

Reviewer 3: ```suggestion
The Kubernetes community [recently announced](/blog/2023/08/31/legacy-package-repository-deprecation/)
```

Author: Fixed :+1:

that their legacy package repositories are frozen and that they will move to the

> [introduced community-owned packages](https://kubernetes.io/blog/2023/08/15/pkgs-k8s-io-introduction) powered by the

Reviewer 3: ```suggestion
[introduced community-owned packages](/blog/2023/08/15/pkgs-k8s-io-introduction) powered by the
```

Author: Fixed :+1:

[openSUSE Build Service (OBS)](https://build.opensuse.org/project/subprojects/isv:kubernetes).
CRI-O has a long history of utilizing [OBS for their package builds](https://github.com/cri-o/cri-o/blob/e292f17/install.md#install-packaged-versions-of-cri-o),
but all of the packaging efforts have been done manually so far.

The CRI-O community absolutely loves Kubernetes, which means that they're
delighted to announce that:

**All future CRI-O packages will be shipped as part of the officially supported
Kubernetes infrastructure hosted on pkgs.k8s.io!**

> There will be a deprecation phase for the existing packages for sure, which is

Reviewer 4: style nit: I think you could drop `for sure` here

Author: Fixed as suggested.

# Install the packages from the added repos
dnf install -y --repo cri-o --repo kubernetes \
    cri-o \
    kubeadm \
    kubectl \
    kubelet

# Enable CRI-O
systemctl enable --now crio.service
```

For `deb` based distributions, they can run (as `root` user):

```bash
# Install dependencies for adding the repositories
apt-get update
apt-get install -y software-properties-common curl

# Add the Kubernetes repository
curl -fsSL https://pkgs.k8s.io/core:/stable:/v1.28/deb/Release.key |
    gpg --dearmor -o /etc/apt/keyrings/kubernetes-apt-keyring.gpg
echo "deb [signed-by=/etc/apt/keyrings/kubernetes-apt-keyring.gpg] https://pkgs.k8s.io/core:/stable:/v1.28/deb/ /" |
    tee /etc/apt/sources.list.d/kubernetes.list

# Add the CRI-O repository
curl -fsSL https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/deb/Release.key |
    gpg --dearmor -o /etc/apt/keyrings/cri-o-apt-keyring.gpg
echo "deb [signed-by=/etc/apt/keyrings/cri-o-apt-keyring.gpg] https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/deb/ /" |
    tee /etc/apt/sources.list.d/cri-o.list

# Install the packages
apt-get update
apt-get install -y cri-o kubelet kubeadm kubectl

# Enable CRI-O
systemctl enable --now crio.service

> ```

Reviewer 5: Same concerns to split and change into a list if it is a sequencial procedure.

init`, which automatically detects that CRI-O is running in the background.
There are also `Vagrantfile` examples available for [Fedora 38](https://github.com/cri-o/packaging/blob/6cdf954/test/fedora/Vagrantfile)
as well as [Ubuntu 22.04](https://github.com/cri-o/packaging/blob/6cdf954/test/ubuntu/Vagrantfile)
for testing the packages together with `kubeadm`.

## How it works under the hood

Everything related to those packages is living in the new [CRI-O
packaging](https://github.com/cri-o/packaging) repository. It contains a [daily reconcilation](https://github.com/cri-o/packaging/actions/workflows/schedule.yml)
GitHub action workflow for all supported release branches as well as tags of
CRI-O. A [test pipeline](https://github.com/cri-o/packaging/actions/workflows/obs.yml)
in the OBS workflow ensures that the packages can be correctly installed and
used before being published. All of the staging and publishing of the
packages is done with the help of the [The Kubernetes Release Toolbox (krel)](https://github.com/kubernetes/release/blob/master/docs/krel/README.md),
which is also used for the official Kubernetes `deb` and `rpm` packages.

> The package build inputs will be reonciled on a daily basis and are provided by

Reviewer 4: reconciled*

CRI-O has a long history of utilizing
[OBS for their package builds](https://github.com/cri-o/cri-o/blob/e292f17/install.md#install-packaged-versions-of-cri-o),
but all of the packaging efforts have been done manually so far.

The CRI-O community absolutely loves Kubernetes, which means that they're
delighted to announce that:

**All future CRI-O packages will be shipped as part of the officially supported
Kubernetes infrastructure hosted on pkgs.k8s.io!**

There will be a deprecation phase for the existing packages, which is currently
being [discussed in the CRI-O community](https://github.com/cri-o/cri-o/discussions/7315).
The new infrastructure will only support releases of CRI-O `>= v1.28.2` as well as
release branches newer than `release-1.28`.

## How to use the new packages

In the same way as the Kubernetes community, CRI-O provides `deb` and `rpm`
packages as part of a dedicated subproject in OBS, called
[`isv:kubernetes:addons:cri-o`](https://build.opensuse.org/project/show/isv:kubernetes:addons:cri-o).
This project acts as an umbrella and provides `stable` (for CRI-O tags) as well as

> `prerelease` (for CRI-O `releaes-1.y` and `main` branches) package builds.

Reviewer: ```suggestion
`prerelease` (for CRI-O `releases-1.y` and `main` branches) package builds.
```

For example, If an end-user would like to install the latest available version
of the CRI-O `main` branch, then they can add the repository in the same way as
they do for Kubernetes.

### `rpm` Based Distributions

For `rpm` based distributions, you can run the following commands as a `root` user
to install CRI-O:

#### Add the Kubernetes repo

```bash
cat <<EOF | tee /etc/yum.repos.d/kubernetes.repo
[kubernetes]
name=Kubernetes
baseurl=https://pkgs.k8s.io/core:/stable:/v1.28/rpm/
enabled=1
gpgcheck=1
gpgkey=https://pkgs.k8s.io/core:/stable:/v1.28/rpm/repodata/repomd.xml.key
EOF
```

#### Add the CRI-O repo

```bash
cat <<EOF | tee /etc/yum.repos.d/cri-o.repo
[cri-o]
name=CRI-O
baseurl=https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/rpm/
enabled=1
gpgcheck=1
gpgkey=https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/rpm/repodata/repomd.xml.key
EOF
```

#### Install official package dependencies

```bash
dnf install -y \
    conntrack \
    container-selinux \
    ebtables \
    ethtool \
    iproute \
    iptables \
    socat
```

#### Install the packages from the added repos

```bash
dnf install -y --repo cri-o --repo kubernetes \
    cri-o \
    kubeadm \
    kubectl \

>     kubelet

Reviewer 6: I would drop this part and later on link to the official instructions. There are some other considerations that are not covered by this blog post, like binding package versions.

Reviewer 6: Okay, that makes sense 👍

```bash
apt-get update
apt-get install -y software-properties-common curl
```

#### Add the Kubernetes repository

```bash
curl -fsSL https://pkgs.k8s.io/core:/stable:/v1.28/deb/Release.key |
    gpg --dearmor -o /etc/apt/keyrings/kubernetes-apt-keyring.gpg
echo "deb [signed-by=/etc/apt/keyrings/kubernetes-apt-keyring.gpg] https://pkgs.k8s.io/core:/stable:/v1.28/deb/ /" |
    tee /etc/apt/sources.list.d/kubernetes.list
```

#### Add the CRI-O repository

```bash
curl -fsSL https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/deb/Release.key |
    gpg --dearmor -o /etc/apt/keyrings/cri-o-apt-keyring.gpg
echo "deb [signed-by=/etc/apt/keyrings/cri-o-apt-keyring.gpg] https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/deb/ /" |
    tee /etc/apt/sources.list.d/cri-o.list
```

#### Install the packages

```bash
apt-get update
apt-get install -y cri-o kubelet kubeadm kubectl
```

#### Start CRI-O

```bash
systemctl start crio.service
```

The Project's `prerelease:/main` prefix at the CRI-O's package path, can be replaced with
`stable:/v1.28`, `stable:/v1.29`, `prerelease:/v1.28` or `prerelease:/v1.29`
if another stream package is used.

> Bootstrapping a cluster using `kubeadm` can be done by running `kubeadm init` command,

Reviewer 6: Let's link this document here: https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/

Author: Done :+1:

CRI-O is running in the background. There are also `Vagrantfile` examples
available for [Fedora 38](https://github.com/cri-o/packaging/blob/main/test/rpm/Vagrantfile)

> as well as [Ubuntu 22.04](https://github.com/cri-o/packaging/blob/main/test/deb/Vagrantfile)

Reviewer: Better to link to a commit ref not a branch; commits are less likely to turn stale.

for testing the packages together with `kubeadm`.

## How it works under the hood

Everything related to these packages lives in the new
[CRI-O packaging repository](https://github.com/cri-o/packaging).

> It contains a [daily reconciliation](https://github.com/crio/packaging/actions/workflows/schedule.yml)

Reviewer: I'd link to the source code of the action and use a commit ref not a branch.

GitHub action workflow, for all supported release branches as well as tags of
CRI-O. A [test pipeline](https://github.com/cri-o/packaging/actions/workflows/obs.yml)
in the OBS workflow ensures that the packages can be correctly installed and
used before being published. All of the staging and publishing of the

> packages is done with the help of the [Kubernetes Release Toolbox (krel)](https://github.com/kubernetes/release/blob/master/docs/krel/README.md),

Reviewer: I'd link to a commit ref not a branch.

which is also used for the official Kubernetes `deb` and `rpm` packages.

The package build inputs will undergo daily reconciliation and will be supplied by

> CRI-O's [static binary bundles](https://console.cloud.google.com/storage/browser/cri-o/artifacts).

Reviewer: ```suggestion
CRI-O's static binary bundles.
```

We shouldn't assume that readers have a Google login.

---
layout: blog
title: "CRI-O is moving towards pkgs.k8s.io"
date: 2023-10-10
slug: cri-o-community-package-infrastructure
---

**Author:** Sascha Grunert

The Kubernetes community [recently announced](/blog/2023/08/31/legacy-package-repository-deprecation/)

> that their legacy package repositories are frozen and that they will move to the

Reviewer 7: ```suggestion
that their legacy package repositories are frozen, and now they will move to 
```

[introduced community-owned packages](/blog/2023/08/15/pkgs-k8s-io-introduction) powered by the
[openSUSE Build Service (OBS)](https://build.opensuse.org/project/subprojects/isv:kubernetes).

> CRI-O has a long history of utilizing [OBS for their package builds](https://github.com/cri-o/cri-o/blob/e292f17/install.md#install-packaged-versions-of-cri-o),
> but all of the packaging efforts have been done manually so far.

Reviewer 7: ```suggestion
CRI-O has a long history of utilizing 
[OBS for their package builds](https://github.com/cri-o/cri-o/blob/e292f17/install.md#install-packaged-versions-of-cri-o), 
but all of the packaging efforts have been done manually so far.
```

The CRI-O community absolutely loves Kubernetes, which means that they're
delighted to announce that:

**All future CRI-O packages will be shipped as part of the officially supported
Kubernetes infrastructure hosted on pkgs.k8s.io!**

There will be a deprecation phase for the existing packages, which is currently
being [discussed in the CRI-O community](https://github.com/cri-o/cri-o/discussions/7315).
The new infrastructure will only support releases of CRI-O `>= v1.28.2` as well as
the corresponding release branches.

## How to use the new packages

> In the same way as the Kubernetes community, CRI-O provides `deb` and `rpm`
> packages as part of a dedicated subproject in OBS, called [isv:kubernetes:addons:cri-o](https://build.opensuse.org/project/show/isv:kubernetes:addons:cri-o).
> This project acts as an umbrella and provides `stable` (for CRI-O tags) as well
> as `prerelease` (for CRI-O `releaes-1.y` and `main` branches) package builds.

Reviewer 7: ```suggestion
In the same way as the Kubernetes community, CRI-O provides `deb` and `rpm`
packages as part of a dedicated subproject in OBS, called 
[isv:kubernetes:addons:cri-o](https://build.opensuse.org/project/show/isv:kubernetes:addons:cri-o). 
This project acts as an umbrella and provides `stable` (for CRI-O tags) as well as 
`prerelease` (for CRI-O `releaes-1.y` and `main` branches) package builds.
```

**Prereleases:**

- [`isv:kubernetes:addons:cri-o:prerelease`](https://build.opensuse.org/project/show/isv:kubernetes:addons:cri-o:prerelease): Prerelease Packages
  - [`isv:kubernetes:addons:cri-o:prerelease:main`](https://build.opensuse.org/project/show/isv:kubernetes:addons:cri-o:prerelease:main): [`main`](https://github.com/cri-o/cri-o/commits/main) branch
  - [`isv:kubernetes:addons:cri-o:prerelease:v1.29`](https://build.opensuse.org/project/show/isv:kubernetes:addons:cri-o:prerelease:v1.29): [`release-1.29`](https://github.com/cri-o/cri-o/commits/release-1.29) branch
  - [`isv:kubernetes:addons:cri-o:prerelease:v1.28`](https://build.opensuse.org/project/show/isv:kubernetes:addons:cri-o:prerelease:v1.28): [`release-1.28`](https://github.com/cri-o/cri-o/commits/release-1.28) branch

There are no stable releases available in the v1.29 repository yet, because
v1.29.0 will be released in December. The CRI-O community will also **not**
support release branches older than `release-1.28`, because there have been CI
requirements merged into `main` which could be only backported to `release-1.28`
with appropriate efforts.

> For example, If an end user would like to install the latest available version

Reviewer 7: ```suggestion
For example, If an end-user would like to install the latest available version
```

of the CRI-O `main` branch, then they can add the repository in the same way as
they do for Kubernetes.

### `rpm` Based Distributions

> For `rpm` based distributions, users can run the following commands as `root`
> to install CRI-O:

Reviewer 7: ```suggestion
For `rpm` based distributions, you can run the following commands as a `root` user 
to install CRI-O:
```

#### Add the Kubernetes repo

```bash
cat <<EOF | tee /etc/yum.repos.d/kubernetes.repo
[kubernetes]
name=Kubernetes
baseurl=https://pkgs.k8s.io/core:/stable:/v1.28/rpm/
enabled=1
gpgcheck=1
gpgkey=https://pkgs.k8s.io/core:/stable:/v1.28/rpm/repodata/repomd.xml.key
EOF
```

#### Add the CRI-O repo

```bash
cat <<EOF | tee /etc/yum.repos.d/cri-o.repo
[cri-o]
name=CRI-O
baseurl=https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/rpm/
enabled=1
gpgcheck=1
gpgkey=https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/rpm/repodata/repomd.xml.key
EOF
```

#### Install official package dependencies

```bash
dnf install -y \
    conntrack \
    container-selinux \
    ebtables \
    ethtool \
    iproute \
    iptables \
    socat
```

#### Install the packages from the added repos

```bash
dnf install -y --repo cri-o --repo kubernetes \
    cri-o \
    kubeadm \
    kubectl \
    kubelet
```

#### Enable and start CRI-O

```bash
systemctl enable --now crio.service
```

### `deb` Based Distributions

> For `deb` based distributions, they can run as `root` user:

Reviewer 7: ```suggestion
For `deb` based distributions, you can run it as a `root` user:
```

#### Install dependencies for adding the repositories

```bash
apt-get update
apt-get install -y software-properties-common curl
```

#### Add the Kubernetes repository

```bash
curl -fsSL https://pkgs.k8s.io/core:/stable:/v1.28/deb/Release.key |
    gpg --dearmor -o /etc/apt/keyrings/kubernetes-apt-keyring.gpg
echo "deb [signed-by=/etc/apt/keyrings/kubernetes-apt-keyring.gpg] https://pkgs.k8s.io/core:/stable:/v1.28/deb/ /" |
    tee /etc/apt/sources.list.d/kubernetes.list
```

#### Add the CRI-O repository

```bash
curl -fsSL https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/deb/Release.key |
    gpg --dearmor -o /etc/apt/keyrings/cri-o-apt-keyring.gpg
echo "deb [signed-by=/etc/apt/keyrings/cri-o-apt-keyring.gpg] https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/deb/ /" |
    tee /etc/apt/sources.list.d/cri-o.list
```

#### Install the packages

```bash
apt-get update
apt-get install -y cri-o kubelet kubeadm kubectl
```

> #### Enable and start CRI-O

Reviewer 7: ```suggestion
#### Start and enable CRI-O
```

```bash
systemctl enable --now crio.service
```

> The `prerelease:/main` project prefix in the CRI-O package paths can be replaced
> with `stable:/v1.28`, `stable:/v1.29`, `prerelease:/v1.28` or `prerelease:/v1.29`
> if another package stream should be used.

Reviewer 7: ```suggestion
The Project's `prerelease:/main` prefix at the CRI-O's package path, can be replaced with 
`stable:/v1.28`, `stable:/v1.29`, `prerelease:/v1.28` or `prerelease:/v1.29`
if another package stream should be used.
```
Line 162: Clarify the meaning, did you wanted to say 
"If another package, stream should be used" or "If another stream package is used."

> Bootstrapping a cluster using `kubeadm` can be then done by running `kubeadm
> init`, which automatically detects that CRI-O is running in the background.

Reviewer 7: ```suggestion
Bootstrapping a cluster using `kubeadm` can be done by running `kubeadm init` command, 
which automatically detects that CRI-O is running in the background.
```

There are also `Vagrantfile` examples available for [Fedora 38](https://github.com/cri-o/packaging/blob/main/test/rpm/Vagrantfile)
as well as [Ubuntu 22.04](https://github.com/cri-o/packaging/blob/main/test/deb/Vagrantfile)
for testing the packages together with `kubeadm`.

## How it works under the hood

> Everything related to those packages is living in the new [CRI-O
> packaging](https://github.com/cri-o/packaging) repository. It contains a [daily reconciliation](https://github.com/cri-o/packaging/actions/workflows/schedule.yml)
> GitHub action workflow for all supported release branches as well as tags of

Reviewer 7: ```suggestion
Everything related to these packages lives in the new 
[CRI-O packaging repository](https://github.com/cri-o/packaging). 
It contains a [daily reconciliation](https://github.com/crio/packaging/actions/workflows/schedule.yml) 
GitHub action workflow, for all supported release branches as well as tags of
```

CRI-O. A [test pipeline](https://github.com/cri-o/packaging/actions/workflows/obs.yml)
in the OBS workflow ensures that the packages can be correctly installed and
used before being published. All of the staging and publishing of the
packages is done with the help of the [The Kubernetes Release Toolbox (krel)](https://github.com/kubernetes/release/blob/master/docs/krel/README.md),
which is also used for the official Kubernetes `deb` and `rpm` packages.

> The package build inputs will be reconciled on a daily basis and are provided by
> the CRI-O [static binary bundles](https://console.cloud.google.com/storage/browser/cri-o/artifacts).
> Those bundles contain everything CRI-O needs to run on a certain architecture
> and are built and signed for each commit in the CRI-O CI. The static builds are
> reproducible, powered by [nixpkgs](https://github.com/NixOS/nixpkgs) and are
> only available for the architectures `x86_64`, `aarch64` and `ppc64le`.

Reviewer 7: ```suggestion
The package build inputs will undergo daily reconciliation and will be supplied by 
CRI-O's [static binary bundles](https://console.cloud.google.com/storage/browser/cri-o/artifacts). 
These bundles are built and signed for each commit in the CRI-O CI, 
and contain everything CRI-O requires to run on a certain architecture. 
The static builds are reproducible, powered by [nixpkgs](https://github.com/NixOS/nixpkgs) 
and available only for `x86_64`, `aarch64` and `ppc64le` architecture.
```

> The CRI-O maintainers are happy for any feedback or suggestion on the new
> packaging efforts! Thank you for reading this blog post, feel free to reach out
> to the maintainers via [the Kubernetes Slack #crio](https://kubernetes.slack.com/messages/CAZH62UR1)
> or the [creating an issue in the packaging repository](https://github.com/cri-o/packaging/issues).

Reviewer 7: ```suggestion
The CRI-O maintainers will be happy to listen to any feedback or suggestions on the new
packaging efforts! Thank you for reading this blog post, feel free to reach out
to the maintainers via the Kubernetes [Slack channel #crio](https://kubernetes.slack.com/messages/CAZH62UR1)
or create an issue in the [packaging repository](https://github.com/cri-o/packaging/issues).
```

---
layout: blog
title: "CRI-O is moving towards pkgs.k8s.io"
date: 2023-10-10
slug: cri-o-community-package-infrastructure
---

**Author:** Sascha Grunert

The Kubernetes community [recently announced](/blog/2023/08/31/legacy-package-repository-deprecation/)

> that their legacy package repositories are frozen, and now they will move to

Reviewer 6: > and now they will move to

I would change this part because we already moved to the new infrastructure.

> [introduced community-owned packages](/blog/2023/08/15/pkgs-k8s-io-introduction) powered by the

Reviewer 6: ```suggestion
[introduced community-owned package repositories](/blog/2023/08/15/pkgs-k8s-io-introduction) powered by the
```

> [openSUSE Build Service (OBS)](https://build.opensuse.org/project/subprojects/isv:kubernetes).

Reviewer 6: ```suggestion
[OpenBuildService (OBS)](https://build.opensuse.org/project/subprojects/isv:kubernetes).
```

CRI-O has a long history of utilizing
[OBS for their package builds](https://github.com/cri-o/cri-o/blob/e292f17/install.md#install-packaged-versions-of-cri-o),
but all of the packaging efforts have been done manually so far.

The CRI-O community absolutely loves Kubernetes, which means that they're
delighted to announce that:

**All future CRI-O packages will be shipped as part of the officially supported
Kubernetes infrastructure hosted on pkgs.k8s.io!**

There will be a deprecation phase for the existing packages, which is currently
being [discussed in the CRI-O community](https://github.com/cri-o/cri-o/discussions/7315).
The new infrastructure will only support releases of CRI-O `>= v1.28.2` as well as

> the corresponding release branches.

Reviewer 6: > as well as the corresponding release branches.

This is slightly confusing to me.

Author: Rephrased to mention the actual release branch.

## How to use the new packages

In the same way as the Kubernetes community, CRI-O provides `deb` and `rpm`
packages as part of a dedicated subproject in OBS, called

> [isv:kubernetes:addons:cri-o](https://build.opensuse.org/project/show/isv:kubernetes:addons:cri-o).

Reviewer 6: ```suggestion
[`isv:kubernetes:addons:cri-o`](https://build.opensuse.org/project/show/isv:kubernetes:addons:cri-o).
```

  - [`isv:kubernetes:addons:cri-o:prerelease:v1.28`](https://build.opensuse.org/project/show/isv:kubernetes:addons:cri-o:prerelease:v1.28): [`release-1.28`](https://github.com/cri-o/cri-o/commits/release-1.28) branch

There are no stable releases available in the v1.29 repository yet, because
v1.29.0 will be released in December. The CRI-O community will also **not**
support release branches older than `release-1.28`, because there have been CI
requirements merged into `main` which could be only backported to `release-1.28`
with appropriate efforts.

For example, If an end-user would like to install the latest available version
of the CRI-O `main` branch, then they can add the repository in the same way as
they do for Kubernetes.

### `rpm` Based Distributions

For `rpm` based distributions, you can run the following commands as a `root` user
to install CRI-O:

#### Add the Kubernetes repo

```bash
cat <<EOF | tee /etc/yum.repos.d/kubernetes.repo
[kubernetes]
name=Kubernetes
baseurl=https://pkgs.k8s.io/core:/stable:/v1.28/rpm/
enabled=1
gpgcheck=1
gpgkey=https://pkgs.k8s.io/core:/stable:/v1.28/rpm/repodata/repomd.xml.key
EOF

> ```

Reviewer 6: We can potentially drop this because you can use CRI-O repos without Kubernetes repos. Instead, we can link to the official installation docs.

Author: We require `kubernetes-cni` in the CRI-O spec.

Reviewer 6: Can we add this as a note? Something like: "we need the kubernetes repo because CRI-O depends on the `kubernetes-cni` package hosted in that repo"

#### Add the CRI-O repo

```bash
cat <<EOF | tee /etc/yum.repos.d/cri-o.repo
[cri-o]
name=CRI-O
baseurl=https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/rpm/
enabled=1
gpgcheck=1
gpgkey=https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/rpm/repodata/repomd.xml.key
EOF
```

#### Install official package dependencies

```bash
dnf install -y \
    conntrack \
    container-selinux \
    ebtables \
    ethtool \
    iproute \
    iptables \
    socat
```

#### Install the packages from the added repos

```bash
dnf install -y --repo cri-o --repo kubernetes \
    cri-o \
    kubeadm \
    kubectl \
    kubelet
```

#### Enable and start CRI-O

```bash
systemctl enable --now crio.service
```

### `deb` Based Distributions

> For `deb` based distributions, you can run it as a `root` user:

Reviewer 6: ```suggestion
For `deb` based distributions, you can run the following commands as a `root` user:
```

Author: Fixed

#### Install dependencies for adding the repositories

```bash
apt-get update
apt-get install -y software-properties-common curl
```

#### Add the Kubernetes repository

```bash
curl -fsSL https://pkgs.k8s.io/core:/stable:/v1.28/deb/Release.key |
    gpg --dearmor -o /etc/apt/keyrings/kubernetes-apt-keyring.gpg
echo "deb [signed-by=/etc/apt/keyrings/kubernetes-apt-keyring.gpg] https://pkgs.k8s.io/core:/stable:/v1.28/deb/ /" |
    tee /etc/apt/sources.list.d/kubernetes.list

> ```

Reviewer 6: We can potentially drop this because you can use CRI-O repos without Kubernetes repos. Instead, we can link to the official installation docs.

Author: I'd be in favor of having the docs how to setup k8s with CRI-O in one place. We can also follow-up and update the kubeadm docs at some later point in time.

#### Add the CRI-O repository

```bash
curl -fsSL https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/deb/Release.key |
    gpg --dearmor -o /etc/apt/keyrings/cri-o-apt-keyring.gpg
echo "deb [signed-by=/etc/apt/keyrings/cri-o-apt-keyring.gpg] https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/deb/ /" |
    tee /etc/apt/sources.list.d/cri-o.list
```

#### Install the packages

```bash
apt-get update
apt-get install -y cri-o kubelet kubeadm kubectl
```

#### Start and enable CRI-O

```bash
systemctl enable --now crio.service
```

The Project's `prerelease:/main` prefix at the CRI-O's package path, can be replaced with
`stable:/v1.28`, `stable:/v1.29`, `prerelease:/v1.28` or `prerelease:/v1.29`

> if another package stream should be used.

Reviewer 7: ```suggestion
if another stream package is used.
```

> if another package stream should be used.

Reviewer 7: Line 164: Clarification of this sentence is required, did you mean

`if another package, stream should be used` (in this case, you are missing a comma `,`)
or 
`if another stream package is used`

Author: The second one. :+1:

Bootstrapping a cluster using `kubeadm` can be done by running `kubeadm init` command, 
which automatically detects that CRI-O is running in the background.
There are also `Vagrantfile` examples available for [Fedora 38](https://github.com/cri-o/packaging/blob/main/test/rpm/Vagrantfile)
as well as [Ubuntu 22.04](https://github.com/cri-o/packaging/blob/main/test/deb/Vagrantfile)
for testing the packages together with `kubeadm`.

## How it works under the hood

Everything related to these packages lives in the new 
[CRI-O packaging repository](https://github.com/cri-o/packaging). 
It contains a [daily reconciliation](https://github.com/crio/packaging/actions/workflows/schedule.yml) 
GitHub action workflow, for all supported release branches as well as tags of
CRI-O. A [test pipeline](https://github.com/cri-o/packaging/actions/workflows/obs.yml)
in the OBS workflow ensures that the packages can be correctly installed and
used before being published. All of the staging and publishing of the

> packages is done with the help of the [The Kubernetes Release Toolbox (krel)](https://github.com/kubernetes/release/blob/master/docs/krel/README.md),

Reviewer 7: ```suggestion
packages is done with the help of the Kubernetes Release [Toolbox (krel)](https://github.com/kubernetes/release/blob/master/docs/krel/README.md),
```

Author: Fixed :+1:
